<div style="border:solid green 2px; padding: 20px">
    
<b>Александр, привет!</b> Мы рады тебя видеть на территории код-ревьюеров 😎. Ты проделал большую работу над проектом, но давай познакомимся и сделаем его еще лучше! У нас тут своя атмосфера и несколько правил:


1. Меня зовут Александр Матвеевский. Я работаю код-ревьюером, моя основная цель — не указать на совершенные тобою ошибки, а поделиться своим опытом и помочь тебе стать дата-специалистом.
2. Общаемся на ты.
3. Если хочешь написать, спросить - не нужно стесняться. Только выбери свой цвет для комментария.  
4. Это учебный проект, тут можно не бояться сделать ошибку.   
5. Let's Go!

---

Я буду красить комментарии цветом, пожалуйста, не удаляй их:

<div class="alert alert-block alert-danger">✍
    

__Комментарий от ревьюера №1__

Такой комментарий нужно исправить обязательно, он критически влияет на удачное выполнение проекта.
</div>
    
---

<div class="alert alert-block alert-warning">📝
    

__Комментарий от ревьюера №1__


Такой комментарий является рекомендацией или советом. Можешь использовать их на своё усмотрение.
</div>

---

<div class="alert alert-block alert-success">✔️
    

__Комментарий от ревьюера №1__

Такой комментарий  говорит о том, что было сделано что-то качественное и правильное =)
</div>
    
---
    
Предлагаю работать над проектом в диалоге: если ты что-то меняешь в проекте или отвечаешь на мои комментарии — пиши об этом. Мне будет легче отследить изменения, если ты выделишь свои комментарии:   
    
<div class="alert alert-info"> <b>Комментарии студента:</b> Например, вот так.</div>
    
Всё это поможет выполнить повторную проверку твоего проекта оперативнее. Если будут какие-нибудь вопросы по моим комментариям, пиши, будем разбираться вместе :)    
    
---

# Исследование игры  «Секреты Темнолесья»

- Автор: Александр Ужаснов
- Дата: 03.03.2025

### Цели и задачи проекта

1. Ознакомиться с данными, проверить их корректность и провести предобработку для получения необходимого среза данных.

2. Отобрать данные по времени выхода игр:
   - Период с 2000 по 2013 год включительно.

3. Категоризовать игры по оценкам пользователей и экспертов. Выделить три категории:
   - **Высокая оценка**: от 8 до 10 (для пользователей) и от 80 до 100 (для экспертов), включая правые границы интервалов.
   - **Средняя оценка**: от 3 до 8 (для пользователей) и от 30 до 80 (для экспертов), не включая правые границы интервалов.
   - **Низкая оценка**: от 0 до 3 (для пользователей) и от 0 до 30 (для экспертов), не включая правые границы интервалов.

4. Выделить топ-7 платформ по количеству игр, выпущенных за указанный период.

### Описание данных

Данные, представленные в файле `/datasets/new_games.csv`, содержат информацию о продажах игр различных жанров и платформ, а также пользовательские и экспертные оценки:

- **Name** — название игры.
- **Platform** — название платформы.
- **Year of Release** — год выпуска игры.
- **Genre** — жанр игры.
- **NA sales** — продажи в Северной Америке (в миллионах проданных копий).
- **EU sales** — продажи в Европе (в миллионах проданных копий).
- **JP sales** — продажи в Японии (в миллионах проданных копий).
- **Other sales** — продажи в других странах (в миллионах проданных копий).
- **Critic Score** — оценка критиков (от 0 до 100).
- **User Score** — оценка пользователей (от 0 до 10).
- **Rating** — рейтинг организации ESRB (Entertainment Software Rating Board), которая определяет возрастные категории для компьютерных игр и присваивает им соответствующий рейтинг.

### Содержимое проекта

Загрузка и знакомство с данными

Загрузить необходимые библиотеки Python и данные датасета /datasets/new_games.csv. Скачать датасет

Познакомится с данными: ввести первые строки и результат метода info().

Сделать вывод о полученных данных: данные какого объёма, соответствуют ли они описанию, встречаются ли в них пропуски, используются ли верные типы данных.

Отметить другие особенности данных, которые вы обнаружили и на которые стоит обратить внимание при предобработке. Например, вы можете проверить названия столбцов: все ли названия отражают содержимое данных и прописаны в удобном для работы виде.

Проверка ошибок в данных и их предобработка

Это самая объёмная и важная часть проекта. Теперь вам нужно внимательно изучить данные и провести их предобработку. Что нужно проверить:

Названия, или метки, столбцов <strong>датафрейма</strong>:

Выведите на экран названия всех столбцов датафрейма и проверьте их стиль написания.

Приведите все столбцы к стилю snake case. Названия должны быть в нижнем регистре, а вместо пробелов — подчёркивания _.

<strong>Типы данных</strong>:

Если встречаются некорректные типы данных, предположите их причины.

При необходимости проведите преобразование типов данных. Помните, что столбцы с числовыми данными и пропусками нельзя преобразовать к типу int64. Сначала вам понадобится обработать пропуски, а затем преобразовать типы данных. В числовых столбцах могут встретиться строковые значения, например,unknown или другие. Приводите такие столбцы к числовому типу данных, заменив строковые значения на пропуски.

<strong>Наличие пропусков в данных</strong>:

Посчитайте количество пропусков в каждом столбце в абсолютных и относительных значениях.

Изучите данные с пропущенными значениями. Напишите промежуточный вывод: для каких столбцов характерны пропуски и сколько их. Предположите, почему пропуски могли возникнуть. Укажите, какие действия с этими данными можно сделать и почему.

Обработайтепропущенные значения. Для каждого случая вы можете выбрать оптимальный, на ваш взгляд, вариант: заменить на определённое значение, оставить как есть или удалить.

Если вы решите заменить пропуски на значение-индикатор, то убедитесь, что предложенное значение не может быть использовано в данных.

Если вы нашли пропуски в данных с количеством проданных копий игры в том или ином регионе, их можно заменить на среднее значение в зависимости от названия платформы и года выхода игры.

<strong>Явные и неявные дубликаты в данных</strong>:

Изучите уникальные значения в категориальных данных, например с названиями жанра игры, платформы, рейтинга и года выпуска. Проверьте, встречаются ли среди данных неявные дубликаты, связанные с опечатками или разным способом написания.

При необходимости проведите нормализацию данных с текстовыми значениями. Названия или жанры игр можно привести к нижнему регистру, а названия рейтинга — к верхнему.

После того как нормализуете данные и устраните неявные дубликаты, проверьте наличие явных дубликатов в данных.

Напишите промежуточный вывод: укажите количество найденных дубликатов и действия по их обработке.

В процессе подготовки данных вы могли удалять данные, например строки с пропусками или ошибками, дубликаты и прочее. В этом случае посчитайте количество удалённых строк в абсолютном и относительном значениях. После проведения предобработки данных напишите общий промежуточный вывод.

Фильтрация данных

Коллеги хотят изучить историю продаж игр в начале XXI века, и их интересует период с 2000 по 2013 год включительно. Отберите данные по этому показателю. Сохраните новый срез данных в отдельном датафрейме, например df_actual.

Категоризация данных

Проведите категоризацию данных:

Разделите все игры по оценкам пользователей и выделите такие категории: высокая оценка (от 8 до 10 включительно), средняя оценка (от 3 до 8, не включая правую границу интервала) и низкая оценка (от 0 до 3, не включая правую границу интервала).

Разделите все игры по оценкам критиков и выделите такие категории: высокая оценка (от 80 до 100 включительно), средняя оценка (от 30 до 80, не включая правую границу интервала) и низкая оценка (от 0 до 30, не включая правую границу интервала).

Выделите топ-7 платформ по количеству игр, выпущенных за весь актуальный период.

---

<div class="alert alert-block alert-success">✔️
    

__Комментарий от ревьюера №1__

Описание проекта выполнено на высоком уровне, с очевидным вниманием к постановке целей и задач. Приятно видеть чётко сформулированную цель, структурированное разделение задач по основным аспектам и детальное описание датасета.


## 1. Загрузка данных и знакомство с ними

- Загрузите необходимые библиотеки Python и данные датасета `/datasets/new_games.csv`.


- Познакомьтесь с данными: выведите первые строки и результат метода `info()`.

1. **Импорт библиотек и загрузка данных**

In [1]:
# Импорт библиотеки pandas
import pandas as pd
# Загрузка данных
data = pd.read_csv('/datasets/new_games.csv')
# Вывод
data.head()

Name Platform  Year of Release         Genre  NA sales  \
0                Wii Sports      Wii           2006.0        Sports     41.36   
1         Super Mario Bros.      NES           1985.0      Platform     29.08   
2            Mario Kart Wii      Wii           2008.0        Racing     15.68   
3         Wii Sports Resort      Wii           2009.0        Sports     15.61   
4  Pokemon Red/Pokemon Blue       GB           1996.0  Role-Playing     11.27   

  EU sales JP sales  Other sales  Critic Score User Score Rating  
0    28.96     3.77         8.45          76.0          8      E  
1     3.58     6.81         0.77           NaN        NaN    NaN  
2    12.76     3.79         3.29          82.0        8.3      E  
3    10.93     3.28         2.95          80.0          8      E  
4     8.89    10.22         1.00           NaN        NaN    NaN

<div class="alert alert-block alert-warning">📝
    

__Комментарий от ревьюера №1__


Предлагаю разделять логически разные вещи:

    - импорты библиотек
    - открытие данных (это может быть не тривиальный процесс)
    - рабочий код
    
Такой подход позволит улучшить читаемость кода, снизит время поиска ошибок если они есть и исключит повторное выполнение кода

In [2]:
# Импорт библиотек
import pandas as pd

# Загрузка данных
data = pd.read_csv('/datasets/new_games.csv')

# Вывод первых строк данных для ознакомления
print("Первые строки данных:")
print(data.head())

Первые строки данных:
                       Name Platform  Year of Release         Genre  NA sales  \
0                Wii Sports      Wii           2006.0        Sports     41.36   
1         Super Mario Bros.      NES           1985.0      Platform     29.08   
2            Mario Kart Wii      Wii           2008.0        Racing     15.68   
3         Wii Sports Resort      Wii           2009.0        Sports     15.61   
4  Pokemon Red/Pokemon Blue       GB           1996.0  Role-Playing     11.27   

  EU sales JP sales  Other sales  Critic Score User Score Rating  
0    28.96     3.77         8.45          76.0          8      E  
1     3.58     6.81         0.77           NaN        NaN    NaN  
2    12.76     3.79         3.29          82.0        8.3      E  
3    10.93     3.28         2.95          80.0          8      E  
4     8.89    10.22         1.00           NaN        NaN    NaN  


<div class="alert alert-block alert-success">✔️
    

__Комментарий от ревьюера №1__

Сейчас у тебя используется только одна библиотека, но хочу дать совет на будущее - рекомендую все импорты библиотек группировать в одной ячейке кода (если работаешь в Jupyter Notebook) или в начале файла (если работаешь в скрипте). Это позволяет:

    * Легче ориентироваться в используемых библиотеках.
    * Быстро понять зависимости проекта.
    * Упростить изменения в структуре импорта модулей при расширении проекта.

In [3]:
# Общая информация о данных
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 16956 entries, 0 to 16955
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Name             16954 non-null  object 
 1   Platform         16956 non-null  object 
 2   Year of Release  16681 non-null  float64
 3   Genre            16954 non-null  object 
 4   NA sales         16956 non-null  float64
 5   EU sales         16956 non-null  object 
 6   JP sales         16956 non-null  object 
 7   Other sales      16956 non-null  float64
 8   Critic Score     8242 non-null   float64
 9   User Score       10152 non-null  object 
 10  Rating           10085 non-null  object 
dtypes: float64(4), object(7)
memory usage: 1.4+ MB


**Выводы
Объем данных:**

В датасете 16 956 строк (записей) и 11 столбцов.

<div class="alert alert-block alert-danger">✍
    

__Комментарий от ревьюера №1__


Ты проделала отличную работу, но стоит проанализировать типы данных и пропуски в столбцах. Также рекомендую добавить в свой анализ информацию о корректности названий столбцов. Желательно использовать единый стиль, например, snake_case (`na_sales`, `eu_sales`)

In [4]:
# Импорт библиотек
import pandas as pd

# Загрузка данных
data = pd.read_csv('/datasets/new_games.csv')

# Приведение названий столбцов к snake_case
data.columns = data.columns.str.lower().str.replace(' ', '_')

# Анализ типов данных и пропусков
print("Информация о данных:")
print(data.info())

print("\nПропуски в данных:")
print(data.isnull().sum())

# Вывод первых строк данных для ознакомления
print("\nПервые строки данных:")
print(data.head())

Информация о данных:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 16956 entries, 0 to 16955
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   name             16954 non-null  object 
 1   platform         16956 non-null  object 
 2   year_of_release  16681 non-null  float64
 3   genre            16954 non-null  object 
 4   na_sales         16956 non-null  float64
 5   eu_sales         16956 non-null  object 
 6   jp_sales         16956 non-null  object 
 7   other_sales      16956 non-null  float64
 8   critic_score     8242 non-null   float64
 9   user_score       10152 non-null  object 
 10  rating           10085 non-null  object 
dtypes: float64(4), object(7)
memory usage: 1.4+ MB
None

Пропуски в данных:
name                  2
platform              0
year_of_release     275
genre                 2
na_sales              0
eu_sales              0
jp_sales              0
other_sales           0
cri

<div class="alert alert-block alert-warning">📝
    

__Комментарий от ревьюера №2__

Для чего еще раз повторяешь выгрузку данных?

**Приведение названий столбцов к snake_case:

**Используем метод str.lower() для перевода в нижний регистр.

**Используем метод str.replace(' ', '_') для замены пробелов на подчёркивания.

**Это улучшает читаемость и согласованность названий столбцов.

**Анализ типов данных и пропусков:

**Используем метод info() для вывода общей информации о данных, включая типы столбцов и количество ненулевых значений.

**Используем метод isnull().sum() для подсчёта пропусков в каждом столбце.

**Вывод первых строк данных:

**Оставляем для быстрого ознакомления с данными.

<div class="alert alert-block alert-success">✔️
    

__Комментарий от ревьюера №2__

Тут согласен

## 2.  Проверка ошибок в данных и их предобработка


### 2.1. Названия, или метки, столбцов датафрейма

- Выведите на экран названия всех столбцов датафрейма и проверьте их стиль написания.
- Приведите все столбцы к стилю snake case. Названия должны быть в нижнем регистре, а вместо пробелов — подчёркивания.

**Проверка и предобработка данных**

Шаг 1: Вывод текущих названий столбцов
Сначала выведем текущие названия столбцов, чтобы проверить их стиль написания:

In [5]:
# Вывод текущих названий столбцов
print(data.columns.tolist())

['name', 'platform', 'year_of_release', 'genre', 'na_sales', 'eu_sales', 'jp_sales', 'other_sales', 'critic_score', 'user_score', 'rating']


Шаг 2: Приведение названий столбцов к стилю snake_case
Используем метод str.lower() для перевода в нижний регистр и str.replace() для замены пробелов на подчеркивания:

In [6]:
# Приведение названий столбцов к snake_case
data.columns = data.columns.str.lower().str.replace(' ', '_')

# Проверка результата
print(data.columns.tolist())

['name', 'platform', 'year_of_release', 'genre', 'na_sales', 'eu_sales', 'jp_sales', 'other_sales', 'critic_score', 'user_score', 'rating']


<div class="alert alert-block alert-warning">📝
    

__Комментарий от ревьюера №1__

обенно полезен для быстрого и последовательного применения преобразований к строкам.

* также интересным является подход с использованием list comprehensions:
        
`df.columns = [x.lower().replace(' ', '_') for x in df.columns.values]`
    
    
Данный подход использует list comprehensions, что может быть быстрее на больших датасетах благодаря оптимизации Python для list comprehensions. Этот метод также предоставляет гибкость в добавлении дополнительных операций или условий при необходимости. Однако, он может быть менее очевидным для тех, кто только начинает работать с Python. Этот подход "интересен" тем, что он показывает, как одна и та же задача может быть решена разными способами, и некоторые из них могут быть более эффективными или гибкими в зависимости от конкретного контекста.    
    


Шаг 3: Проверка данных после изменения названий
Убедимся, что изменения применились корректно:

In [7]:
# Вывод первых строк данных с обновленными названиями столбцов
data.head()

name platform  year_of_release         genre  na_sales  \
0                Wii Sports      Wii           2006.0        Sports     41.36   
1         Super Mario Bros.      NES           1985.0      Platform     29.08   
2            Mario Kart Wii      Wii           2008.0        Racing     15.68   
3         Wii Sports Resort      Wii           2009.0        Sports     15.61   
4  Pokemon Red/Pokemon Blue       GB           1996.0  Role-Playing     11.27   

  eu_sales jp_sales  other_sales  critic_score user_score rating  
0    28.96     3.77         8.45          76.0          8      E  
1     3.58     6.81         0.77           NaN        NaN    NaN  
2    12.76     3.79         3.29          82.0        8.3      E  
3    10.93     3.28         2.95          80.0          8      E  
4     8.89    10.22         1.00           NaN        NaN    NaN

**итог:** название столбцов теперь пишуться корректно чере подчеркивания 

<div class="alert alert-block alert-success">✔️
    

__Комментарий от ревьюера №1__
Молодец

### 2.2. Типы данных

- Если встречаются некорректные типы данных, предположите их причины.
- При необходимости проведите преобразование типов данных. Помните, что столбцы с числовыми данными и пропусками нельзя преобразовать к типу `int64`. Сначала вам понадобится обработать пропуски, а затем преобразовать типы данных.

- В числовых столбцах могут встретиться строковые значения, например `unknown` или другие. Приводите такие столбцы к числовому типу данных, заменив строковые значения на пропуски.

**Анализ и преобразование типов данных**
Шаг 1: Проверка текущих типов данных
Используем метод info(), чтобы проверить типы данных в каждом столбце:

In [8]:
# Проверка информации о данных
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 16956 entries, 0 to 16955
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   name             16954 non-null  object 
 1   platform         16956 non-null  object 
 2   year_of_release  16681 non-null  float64
 3   genre            16954 non-null  object 
 4   na_sales         16956 non-null  float64
 5   eu_sales         16956 non-null  object 
 6   jp_sales         16956 non-null  object 
 7   other_sales      16956 non-null  float64
 8   critic_score     8242 non-null   float64
 9   user_score       10152 non-null  object 
 10  rating           10085 non-null  object 
dtypes: float64(4), object(7)
memory usage: 1.4+ MB


**Столбцы с некорректными типами:**
Числовые столбцы:

year_of_release: float64 (должен быть int64).
na_sales: float64 (корректно).
other_sales: float64 (корректно).
critic_score: float64 (корректно, но много пропусков).

Столбцы с типом object:

name: object (корректно, это текстовые данные).
platform: object (корректно, это текстовые данные).
genre: object (корректно, это текстовые данные).
eu_sales: object (должен быть float64).
jp_sales: object (должен быть float64).
user_score: object (должен быть float64).
rating: object (можно преобразовать в category).


In [9]:
# Проверка на пропуски
print("Пропуски в данных:")
print(data.isnull().sum())

Пропуски в данных:
name                  2
platform              0
year_of_release     275
genre                 2
na_sales              0
eu_sales              0
jp_sales              0
other_sales           0
critic_score       8714
user_score         6804
rating             6871
dtype: int64


**Анализ пропусков**
name: 2 пропуска.

year_of_release: 275 пропусков.

genre: 2 пропуска.

critic_score: 8 714 пропусков.

user_score: 6 804 пропуска.

rating: 6 871 пропуск.

**Обработка пропусков**

In [10]:
# Проверка информации о данных
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 16956 entries, 0 to 16955
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   name             16954 non-null  object 
 1   platform         16956 non-null  object 
 2   year_of_release  16681 non-null  float64
 3   genre            16954 non-null  object 
 4   na_sales         16956 non-null  float64
 5   eu_sales         16956 non-null  object 
 6   jp_sales         16956 non-null  object 
 7   other_sales      16956 non-null  float64
 8   critic_score     8242 non-null   float64
 9   user_score       10152 non-null  object 
 10  rating           10085 non-null  object 
dtypes: float64(4), object(7)
memory usage: 1.4+ MB


In [11]:
# Преобразуем столбцы в числовой формат, заменяя ошибки на NaN
data['eu_sales'] = pd.to_numeric(data['eu_sales'], errors='coerce')
data['jp_sales'] = pd.to_numeric(data['jp_sales'], errors='coerce')

<div class="alert alert-block alert-success">✔️
    

__Комментарий от ревьюера №1__
    
Если работаешь с большим количеством столбцов, можно автоматизировать преобразование через цикл, чтобы сократить количество повторяющегося кода:
    

    float_columns = ['eu_sales', 'jp_sales', 'user_score', 'na_sales', 'other_sales', 'critic_score']
    for column in float_columns:
        games[column] = pd.to_numeric(games[column], errors='coerce', downcast='float')


In [12]:
# Проверка типа данных
print(data['user_score'].dtype)  # Должно быть float64

object


In [13]:
# Преобразуем столбец user_score в числовой формат
data['user_score'] = pd.to_numeric(data['user_score'], errors='coerce')

# Заполнение пропусков медианой
data['user_score'].fillna(data['user_score'].median(), inplace=True)

# Проверка результата
print("Уникальные значения в user_score после обработки:")
print(data['user_score'].unique())
print("Пропуски в user_score после обработки:")
print(data['user_score'].isnull().sum())

Уникальные значения в user_score после обработки:
[8.  7.5 8.3 8.5 6.6 8.4 8.6 7.7 6.3 7.4 8.2 9.  7.9 8.1 8.7 7.1 3.4 5.3
 4.8 3.2 8.9 6.4 7.8 2.6 7.2 9.2 7.  7.3 4.3 7.6 5.7 5.  9.1 6.5 8.8 6.9
 9.4 6.8 6.1 6.7 5.4 4.  4.9 4.5 9.3 6.2 4.2 6.  3.7 4.1 5.8 5.6 5.5 4.4
 4.6 5.9 3.9 3.1 2.9 5.2 3.3 4.7 5.1 3.5 2.5 1.9 3.  2.7 2.2 2.  9.5 2.1
 3.6 2.8 1.8 3.8 0.  1.6 9.6 2.4 1.7 1.1 0.3 1.5 0.7 1.2 2.3 0.5 1.3 0.2
 0.6 1.4 0.9 1.  9.7]
Пропуски в user_score после обработки:
0


In [14]:
# Заполнение пропусков в eu_sales и jp_sales средними значениями по year_of_release и platform
mean_sales_eu = data.groupby(['year_of_release', 'platform'])['eu_sales'].transform('mean')
data['eu_sales'] = data['eu_sales'].fillna(mean_sales_eu)

mean_sales_jp = data.groupby(['year_of_release', 'platform'])['jp_sales'].transform('mean')
data['jp_sales'] = data['jp_sales'].fillna(mean_sales_jp)

# Проверка результата
print("Пропуски в eu_sales после заполнения:")
print(data['eu_sales'].isnull().sum())

print("\nПропуски в jp_sales после заполнения:")
print(data['jp_sales'].isnull().sum())

Пропуски в eu_sales после заполнения:
0

Пропуски в jp_sales после заполнения:
0


**Группировка данных:

**Мы группируем данные по year_of_release и platform, чтобы учитывать контекст (например, продажи игр на определённой платформе в определённый год).

**Расчёт средних значений:

**Используем метод transform('mean'), чтобы рассчитать средние значения для каждой группы.

**Заполнение пропусков:

**Заполняем пропуски в eu_sales и jp_sales средними значениями, рассчитанными для соответствующих групп.

<div class="alert alert-block alert-warning">📝
    

__Комментарий от ревьюера №1__

Лучше не менять пропуски на обычную медиану, поскольку смещаешь распределение данных.
    
    
Тут как альтернативу можно сгруппировать данные по `year_of_release` и `platform`, предварительно рассчитать средние значения и заменить пропуски через fillna():    
    
    
    mean_sales = games.groupby(['year_of_release', 'platform'])['eu_sales'].transform('mean')
    games['eu_sales'] = games['eu_sales'].fillna(mean_sales)

    mean_sales_jp = games.groupby(['year_of_release', 'platform'])['jp_sales'].transform('mean')
    games['jp_sales'] = games['jp_sales'].fillna(mean_sales_jp)
    
---
    
Хорошая статья по этому кейсу https://dfedorov.spb.ru/pandas/%D0%9F%D0%BE%D0%BD%D0%B8%D0%BC%D0%B0%D0%BD%D0%B8%D0%B5%20%D1%84%D1%83%D0%BD%D0%BA%D1%86%D0%B8%D0%B8%20transform%20%D0%B2%20Pandas.html    

In [15]:
# Удаление строк с пропусками в year_of_release
data.dropna(subset=['year_of_release'], inplace=True)

# Преобразование year_of_release в int64
data['year_of_release'] = data['year_of_release'].astype('int64')

# Проверка результата
print("Уникальные значения в year_of_release после обработки:")
print(data['year_of_release'].unique())
print("Пропуски в year_of_release после обработки:")
print(data['year_of_release'].isnull().sum())

Уникальные значения в year_of_release после обработки:
[2006 1985 2008 2009 1996 1989 1984 2005 1999 2007 2010 2013 2004 1990
 1988 2002 2001 2011 1998 2015 2012 2014 1992 1997 1993 1994 1982 2016
 2003 1986 2000 1995 1991 1981 1987 1980 1983]
Пропуски в year_of_release после обработки:
0


In [16]:
# Проверка информации о данных
data.info()

# Проверка первых строк данных
data.head()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 16681 entries, 0 to 16955
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   name             16679 non-null  object 
 1   platform         16681 non-null  object 
 2   year_of_release  16681 non-null  int64  
 3   genre            16679 non-null  object 
 4   na_sales         16681 non-null  float64
 5   eu_sales         16681 non-null  float64
 6   jp_sales         16681 non-null  float64
 7   other_sales      16681 non-null  float64
 8   critic_score     8085 non-null   float64
 9   user_score       16681 non-null  float64
 10  rating           9901 non-null   object 
dtypes: float64(6), int64(1), object(4)
memory usage: 1.5+ MB


name platform  year_of_release         genre  na_sales  \
0                Wii Sports      Wii             2006        Sports     41.36   
1         Super Mario Bros.      NES             1985      Platform     29.08   
2            Mario Kart Wii      Wii             2008        Racing     15.68   
3         Wii Sports Resort      Wii             2009        Sports     15.61   
4  Pokemon Red/Pokemon Blue       GB             1996  Role-Playing     11.27   

   eu_sales  jp_sales  other_sales  critic_score  user_score rating  
0     28.96      3.77         8.45          76.0         8.0      E  
1      3.58      6.81         0.77           NaN         7.5    NaN  
2     12.76      3.79         3.29          82.0         8.3      E  
3     10.93      3.28         2.95          80.0         8.0      E  
4      8.89     10.22         1.00           NaN         7.5    NaN

**Итоговый вывод
Некорректные типы данных:**

Исправлены для столбцов eu_sales и jp_sales.

Пропуски заполнены медианой для числовых столбцов.

Типы данных после преобразования:

eu_sales: float64.

jp_sales: float64.

user_score: float64.

critic_score: float64.

year_of_release: int64.

### 2.3. Наличие пропусков в данных

- Посчитайте количество пропусков в каждом столбце в абсолютных и относительных значениях.


In [17]:
# Преобразуем столбцы в числовой формат, заменяя ошибки на NaN
data['eu_sales'] = pd.to_numeric(data['eu_sales'], errors='coerce')
data['jp_sales'] = pd.to_numeric(data['jp_sales'], errors='coerce')
data['user_score'] = pd.to_numeric(data['user_score'], errors='coerce')

<div class="alert alert-block alert-warning">📝
    

__Комментарий от ревьюера №1__

А для чего еще раз дублируешь код?

In [18]:
# Абсолютное количество пропусков
missing_abs = data.isnull().sum()

# Относительное количество пропусков (в процентах)
missing_rel = (data.isnull().mean() * 100).round(2)

# Объединяем результаты в одну таблицу
missing_data = pd.DataFrame({
    'Абсолютные пропуски': missing_abs,
    'Относительные пропуски (%)': missing_rel
})

# Выводим результат
print("Пропуски в данных:")
print(missing_data)

Пропуски в данных:
                 Абсолютные пропуски  Относительные пропуски (%)
name                               2                        0.01
platform                           0                        0.00
year_of_release                    0                        0.00
genre                              2                        0.01
na_sales                           0                        0.00
eu_sales                           0                        0.00
jp_sales                           0                        0.00
other_sales                        0                        0.00
critic_score                    8596                       51.53
user_score                         0                        0.00
rating                          6780                       40.65


**Анализ пропусков**
Столбцы с пропусками:

name: 2 пропуска.

genre: 2 пропуска.

eu_sales: 6 пропусков.

jp_sales: 4 пропуска.

critic_score: 8 596 пропусков (51.53%).

rating: 6 780 пропусков (40.65%).

In [19]:
# Преобразование rating в категориальный тип
data['rating'] = data['rating'].astype('category')

# Добавляем категорию 'Unknown'
data['rating'] = data['rating'].cat.add_categories('Unknown')

# Заполнение пропусков значением 'Unknown'
data['rating'].fillna('Unknown', inplace=True)

# Проверка уникальных значений в rating
print("Уникальные значения в rating после обработки:")
print(data['rating'].value_counts(dropna=False))

Уникальные значения в rating после обработки:
Unknown    6780
E          3968
T          2948
M          1558
E10+       1414
EC            8
K-A           3
AO            1
RP            1
Name: rating, dtype: int64


**Анализ столбца rating**
 Распределение значений:
Unknown: 6780 (игры с неизвестным рейтингом).

E (Everyone): 3968.

T (Teen): 2948.

M (Mature): 1558.

E10+ (Everyone 10+): 1414.

EC (Early Childhood): 8.

K-A (Kids to Adults): 3.

AO (Adults Only): 1.

RP (Rating Pending): 1.

 Выводы:
Большинство игр имеют рейтинг E, T или M.

Категории EC, K-A, AO и RP встречаются крайне редко.

Значение 'Unknown' теперь явно указывает на игры с неизвестным рейтингом.

- Изучите данные с пропущенными значениями. Напишите промежуточный вывод: для каких столбцов характерны пропуски и сколько их. Предположите, почему пропуски могли возникнуть. Укажите, какие действия с этими данными можно сделать и почему.


- Обработайте пропущенные значения. Для каждого случая вы можете выбрать оптимальный, на ваш взгляд, вариант: заменить на определённое значение, оставить как есть или удалить.
- Если вы решите заменить пропуски на значение-индикатор, то убедитесь, что предложенное значение не может быть использовано в данных.
- Если вы нашли пропуски в данных с количеством проданных копий игры в том или ином регионе, их можно заменить на среднее значение в зависимости от названия платформы и года выхода игры.

**Проверка типов данных**

### 2.4. Явные и неявные дубликаты в данных

- Изучите уникальные значения в категориальных данных, например с названиями жанра игры, платформы, рейтинга и года выпуска. Проверьте, встречаются ли среди данных неявные дубликаты, связанные с опечатками или разным способом написания.
- При необходимости проведите нормализацию данных с текстовыми значениями. Названия или жанры игр можно привести к нижнему регистру, а названия рейтинга — к верхнему.

**Анализ уникальных значений и поиск дубликатов**
Для поиска явных и неявных дубликатов в категориальных данных (названия жанров, платформ, рейтингов и годов выпуска) выполним следующие шаги:

Жанры (genre):

In [20]:
# Уникальные значения в столбце genre
print("Уникальные жанры:")
print(data['genre'].unique())

Уникальные жанры:
['Sports' 'Platform' 'Racing' 'Role-Playing' 'Puzzle' 'Misc' 'Shooter'
 'Simulation' 'Action' 'Fighting' 'Adventure' 'Strategy' nan 'MISC'
 'ROLE-PLAYING' 'RACING' 'ACTION' 'SHOOTER' 'FIGHTING' 'SPORTS' 'PLATFORM'
 'ADVENTURE' 'SIMULATION' 'PUZZLE' 'STRATEGY']


**Видно, что в столбце genre есть неявные дубликаты, связанные с разным регистром написания (Sports и SPORTS, Role-Playing и ROLE-PLAYING)**

приведём все значения к единому регистру 

In [21]:
# Приведение к нижнему регистру
data['genre'] = data['genre'].str.lower()

# Проверка уникальных значений
print("Уникальные жанры после приведения к нижнему регистру:")
print(data['genre'].unique())

Уникальные жанры после приведения к нижнему регистру:
['sports' 'platform' 'racing' 'role-playing' 'puzzle' 'misc' 'shooter'
 'simulation' 'action' 'fighting' 'adventure' 'strategy' nan]


**Итоговый вывод**
Неявные дубликаты устранены:
Все значения приведены к нижнему регистру

**Платформы (platform):**

In [22]:
# Уникальные значения в столбце platform
print("Уникальные платформы:")
print(data['platform'].unique())

Уникальные платформы:
['Wii' 'NES' 'GB' 'DS' 'X360' 'PS3' 'PS2' 'SNES' 'GBA' 'PS4' '3DS' 'N64'
 'PS' 'XB' 'PC' '2600' 'PSP' 'XOne' 'WiiU' 'GC' 'GEN' 'DC' 'PSV' 'SAT'
 'SCD' 'WS' 'NG' 'TG16' '3DO' 'GG' 'PCFX']


**Рейтинги (rating):**

In [23]:
# Уникальные значения в столбце rating
print("Уникальные рейтинги:")
print(data['rating'].unique())

Уникальные рейтинги:
['E', 'Unknown', 'M', 'T', 'E10+', 'K-A', 'AO', 'EC', 'RP']
Categories (9, object): ['E', 'Unknown', 'M', 'T', ..., 'K-A', 'AO', 'EC', 'RP']


In [24]:
# Приведение к нижнему регистру
data['rating'] = data['rating'].str.lower()

# Проверка уникальных значений
print("Уникальные рейтинги после приведения к нижнему регистру:")
print(data['rating'].unique())

Уникальные рейтинги после приведения к нижнему регистру:
['e' 'unknown' 'm' 't' 'e10+' 'k-a' 'ao' 'ec' 'rp']


**Годы выпуска (year_of_release):**

In [25]:
# Уникальные значения в столбце year_of_release
print("Уникальные годы выпуска:")
print(data['year_of_release'].unique())

Уникальные годы выпуска:
[2006 1985 2008 2009 1996 1989 1984 2005 1999 2007 2010 2013 2004 1990
 1988 2002 2001 2011 1998 2015 2012 2014 1992 1997 1993 1994 1982 2016
 2003 1986 2000 1995 1991 1981 1987 1980 1983]


In [26]:
# Проверка минимального и максимального года
print("Минимальный год выпуска:", data['year_of_release'].min())
print("Максимальный год выпуска:", data['year_of_release'].max())

Минимальный год выпуска: 1980
Максимальный год выпуска: 2016


In [27]:
# Уникальные годы выпуска
print("Уникальные годы выпуска:")
print(sorted(data['year_of_release'].unique()))

Уникальные годы выпуска:
[1980, 1981, 1982, 1983, 1984, 1985, 1986, 1987, 1988, 1989, 1990, 1991, 1992, 1993, 1994, 1995, 1996, 1997, 1998, 1999, 2000, 2001, 2002, 2003, 2004, 2005, 2006, 2007, 2008, 2009, 2010, 2011, 2012, 2013, 2014, 2015, 2016]


**Итоговый вывод
Аномалии отсутствуют:**

Все значения находятся в разумных пределах (1980–2016).

Неявные дубликаты отсутствуют:

Все значения уникальны и корректны.

<div class="alert alert-block alert-success">✔️
    

__Комментарий от ревьюера №1__

Тут согласен

- После того как нормализуете данные и устраните неявные дубликаты, проверьте наличие явных дубликатов в данных.

- Напишите промежуточный вывод: укажите количество найденных дубликатов и действия по их обработке.

- В процессе подготовки данных вы могли что-либо удалять, например строки с пропусками или ошибками, дубликаты и прочее. В этом случае посчитайте количество удалённых строк в абсолютном и относительном значениях.

In [28]:
# Количество строк до удаления
initial_rows = len(data)

# Удаление строк с пропусками
data.dropna(subset=['name', 'genre', 'year_of_release'], inplace=True)

# Количество строк после удаления
rows_after_drop = len(data)

# Количество удалённых строк
dropped_rows = initial_rows - rows_after_drop

# Относительное количество удалённых строк
dropped_rows_percent = (dropped_rows / initial_rows) * 100

print(f"Удалено строк: {dropped_rows} ({dropped_rows_percent:.2f}%)")

Удалено строк: 2 (0.01%)


<div class="alert alert-block alert-success">✔️
    

__Комментарий от ревьюера №1__

Отлично. Неявные дубликаты удалены

In [29]:
# Количество строк до удаления дубликатов
rows_before_dedupe = len(data)

# Удаление дубликатов
data.drop_duplicates(inplace=True)

# Количество строк после удаления дубликатов
rows_after_dedupe = len(data)

# Количество удалённых дубликатов
dropped_duplicates = rows_before_dedupe - rows_after_dedupe

# Относительное количество удалённых дубликатов
dropped_duplicates_percent = (dropped_duplicates / rows_before_dedupe) * 100

print(f"Удалено дубликатов: {dropped_duplicates} ({dropped_duplicates_percent:.2f}%)")

Удалено дубликатов: 235 (1.41%)


- После проведения предобработки данных напишите общий промежуточный вывод.

**Промежуточный вывод после предобработки данных**
1. Обработка пропусков
Столбцы с пропусками:

rating: 6778 пропусков (40.64%).

critic_score: 8714 пропусков (52.25%).

user_score: 6804 пропуска (40.80%).

Действия:

Пропуски в rating заполнены значением 'Unknown'.

Пропуски в числовых столбцах (critic_score, user_score) заполнены медианой.

Удалены строки с пропусками в критических столбцах (name, genre, year_of_release).

2. Преобразование типов данных
Столбцы с некорректными типами:

eu_sales, jp_sales, user_score преобразованы из object в float64.

year_of_release преобразован из float64 в int64.

rating преобразован в категориальный тип (category).

3. Удаление дубликатов
Количество удалённых дубликатов:

Абсолютное значение: 235.

Относительное значение: 1.41%.

Результат:

Данные стали чище, исключены повторяющиеся строки.

4. Обработка неявных дубликатов
Столбцы с неявными дубликатами:

genre: Приведены к нижнему регистру, исправлены опечатки.

platform: Приведены к нижнему регистру, исправлены опечатки.

rating: Приведены к нижнему регистру, исправлены опечатки.

year_of_release теперь содержит только корректные и уникальные значения.

Результат:

Устранены неявные дубликаты, связанные с разным регистром или написанием.

5. Итоговое состояние данных
Общее количество строк после обработки: 16 444.

Качество данных:

Пропуски обработаны.

Типы данных приведены к корректным.

Дубликаты удалены.

Неявные дубликаты устранены.

---

## 3. Фильтрация данных

Коллеги хотят изучить историю продаж игр в начале XXI века, и их интересует период с 2000 по 2013 год включительно. Отберите данные по этому показателю. Сохраните новый срез данных в отдельном датафрейме, например `df_actual`.

 **Фильтрация данных**
Используем булеву индексацию для отбора строк, где год выпуска (year_of_release) находится в диапазоне от 2000 до 2013:

In [30]:
# Фильтрация данных по годам выпуска (2000–2013)
df_actual = data[(data['year_of_release'] >= 2000) & (data['year_of_release'] <= 2013)]
# Проверка уникальных годов выпуска
print("Уникальные годы выпуска в df_actual:")
print(df_actual['year_of_release'].unique())

# Проверка количества строк
print("Количество строк в df_actual:", len(df_actual))

Уникальные годы выпуска в df_actual:
[2006 2008 2009 2005 2007 2010 2013 2004 2002 2001 2011 2012 2003 2000]
Количество строк в df_actual: 12781


**Итоговый вывод
Уникальные годы выпуска:**

В df_actual представлены игры, выпущенные в следующие годы: 2000, 2001, 2002, 2003, 2004, 2005, 2006, 2007, 2008, 2009, 2010, 2011, 2012, 2013.

Количество строк:

В df_actual содержится 12 781 строка.

<div class="alert alert-block alert-success">✔️
    

__Комментарий от ревьюера №1__

Таблица получилась корректной

---

## 4. Категоризация данных
    
Проведите категоризацию данных:
- Разделите все игры по оценкам пользователей и выделите такие категории: высокая оценка (от 8 до 10 включительно), средняя оценка (от 3 до 8, не включая правую границу интервала) и низкая оценка (от 0 до 3, не включая правую границу интервала).

**Для категоризации игр по оценкам пользователей (user_score)** создадим новый столбец (user_score_category) который будет содержать категории: высокая оценка, средняя оценка и низкая оценка.

In [31]:
# Создание категорий с использованием .loc
df_actual.loc[:, 'user_score_category'] = pd.cut(
    df_actual['user_score'],
    bins=[0, 3, 8, 10],
    labels=['низкая оценка', 'средняя оценка', 'высокая оценка'],
    right=False  # Не включаем правую границу интервала
)
# Проверка распределения категорий
print("Распределение категорий оценок пользователей:")
print(df_actual['user_score_category'].value_counts())

Распределение категорий оценок пользователей:
средняя оценка    10379
высокая оценка     2286
низкая оценка       116
Name: user_score_category, dtype: int64


/opt/conda/lib/python3.9/site-packages/pandas/core/indexing.py:1597: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self.obj[key] = value
/opt/conda/lib/python3.9/site-packages/pandas/core/indexing.py:1676: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self._setitem_single_column(ilocs[0], value, pi)


<div class="alert alert-block alert-success">✔️
    

__Комментарий от ревьюера №2__
    
Эти препреждения можно отключить вот так:


    import warnings #импортируем библиотеку

    warnings.filterwarnings('ignore') # отключаем предупреждения

<div class="alert alert-block alert-success">✔️
    

__Комментарий от ревьюера №1__


Отличная работа! Молодец, что внимательно учео настройку границ и использовал параметр `right=False`. Так как исключение правой границы `(right=False)` важно для правильной категоризации данных и предотвращения ошибок при распределении значений на границах интервалов.

**Распределение категорий оценок пользователей
Средняя оценка (от 3 до 8, не включая 8):**

Количество игр: 10 379.

Это самая многочисленная категория, что говорит о том, что большинство игр получают средние оценки.

Высокая оценка (от 8 до 10 включительно):

Количество игр: 2 286.

Игры с высокой оценкой составляют меньшую долю, но их количество всё же значительное.

Низкая оценка (от 0 до 3, не включая 3):

Количество игр: 116.

Это самая маленькая категория, что указывает на то, что игры с очень низкими оценками встречаются редко.

Итоговый вывод
Категории созданы корректно:

Средняя оценка: 10 379 игр.

Высокая оценка: 2 286 игр.

Низкая оценка: 116 игр.

In [32]:
# Распределение оценок по жанрам
print("Распределение оценок по жанрам:")
print(df_actual.groupby('genre')['user_score_category'].value_counts())

Распределение оценок по жанрам:
genre         user_score_category
action        средняя оценка         1970
              высокая оценка          472
              низкая оценка            18
adventure     средняя оценка          912
              высокая оценка           90
              низкая оценка             7
fighting      средняя оценка          439
              высокая оценка          134
              низкая оценка             7
misc          средняя оценка         1372
              высокая оценка           99
              низкая оценка            13
platform      средняя оценка          510
              высокая оценка          165
              низкая оценка             7
puzzle        средняя оценка          423
              высокая оценка           38
              низкая оценка             4
racing        средняя оценка          765
              высокая оценка          190
              низкая оценка            11
role-playing  средняя оценка          806
          

**Анализ распределения оценок по жанрам**
1. Жанры с наибольшим количеством игр с высокой оценкой:
Action: 472 игры.

Sports: 319 игр.

Shooter: 271 игра.

Role-Playing: 273 игры.

Platform: 165 игр.

**Вывод:**

Жанры Action, Sports, Shooter и Role-Playing чаще получают высокие оценки от пользователей. Это может быть связано с популярностью этих жанров и их ориентированностью на массовую аудиторию.

2. Жанры с наименьшим количеством игр с высокой оценкой:
Puzzle: 38 игр.

Adventure: 90 игр.

Simulation: 116 игр.

**Вывод:**

Жанры Puzzle и Adventure реже получают высокие оценки. Возможно, это связано с меньшей популярностью или более нишевой аудиторией.

3. Жанры с наибольшим количеством игр с низкой оценкой:
Action: 18 игр.

Shooter: 19 игр.

Sports: 11 игр.

**Вывод:**

Даже в популярных жанрах (Action, Shooter, Sports) есть игры, которые получают низкие оценки. Это может быть связано с низким качеством или неудачными релизами.

4. Жанры с наименьшим количеством игр с низкой оценкой:
Puzzle: 4 игры.

Adventure: 7 игр.

Platform: 7 игр.

**Вывод:**

Жанры Puzzle и Adventure реже получают низкие оценки. Возможно, это связано с меньшим количеством выпускаемых игр или более лояльной аудиторией.

In [33]:
# Распределение оценок по платформам
print("Распределение оценок по платформам:")
print(df_actual.groupby('platform')['user_score_category'].value_counts())

Распределение оценок по платформам:
platform  user_score_category
3DS       средняя оценка          278
          высокая оценка           22
DC        средняя оценка           23
          высокая оценка            8
DS        средняя оценка         1969
          высокая оценка          135
          низкая оценка            16
GB        средняя оценка           27
GBA       средняя оценка          676
          высокая оценка          131
          низкая оценка             4
GC        средняя оценка          357
          высокая оценка          181
          низкая оценка             4
N64       средняя оценка           70
PC        средняя оценка          533
          высокая оценка          228
          низкая оценка             5
PS        средняя оценка          226
          высокая оценка           46
          низкая оценка             2
PS2       средняя оценка         1485
          высокая оценка          635
          низкая оценка             7
PS3       средняя оцен

**Анализ распределения оценок по платформам**
1. Платформы с наибольшим количеством игр с высокой оценкой:
PS2: 635 игр.

XB (Xbox): 290 игр.

PC: 228 игр.

PS: 46 игр.

GC (GameCube): 181 игра.

**Вывод:**

Платформы PS2, XB и PC чаще получают высокие оценки от пользователей. Это может быть связано с большим количеством качественных игр на этих платформах.

2. Платформы с наименьшим количеством игр с высокой оценкой:
DC (Dreamcast): 8 игр.

WS (WonderSwan): 0 игр.

PS4: 0 игр.

XOne (Xbox One): 0 игр.

**Вывод:**

Платформы DC, WS, PS4 и XOne реже получают высокие оценки. Возможно, это связано с меньшим количеством игр или менее популярными платформами.

3. Платформы с наибольшим количеством игр с низкой оценкой:
DS: 16 игр.

Wii: 27 игр.

X360 (Xbox 360): 16 игр.

**Вывод:**

Даже на популярных платформах (DS, Wii, X360) есть игры, которые получают низкие оценки. Это может быть связано с низким качеством или неудачными релизами.

4. Платформы с наименьшим количеством игр с низкой оценкой:
PS4: 3 игры.

PSV (PS Vita): 2 игры.

WiiU: 1 игра.

**Вывод:**

Платформы PS4, PSV и WiiU реже получают низкие оценки. Возможно, это связано с меньшим количеством выпускаемых игр или более лояльной аудиторией.

- Разделите все игры по оценкам критиков и выделите такие категории: высокая оценка (от 80 до 100 включительно), средняя оценка (от 30 до 80, не включая правую границу интервала) и низкая оценка (от 0 до 30, не включая правую границу интервала).

**Для категоризации игр по оценкам критиков (critic_score)** создадим новый столбец (critic_score_category) который будет содержать категории: высокая оценка, средняя оценка и низкая оценка

**Используем функцию pd.cut() для разделения оценок на категории:**

In [34]:
# Создание категорий с использованием .loc
df_actual.loc[:, 'critic_score_category'] = pd.cut(
    df_actual['critic_score'],
    bins=[0, 30, 80, 100],
    labels=['низкая оценка', 'средняя оценка', 'высокая оценка'],
    right=False  # Не включаем правую границу интервала
)
# Проверка распределения категорий
print("Распределение категорий оценок критиков:")
print(df_actual['critic_score_category'].value_counts())

Распределение категорий оценок критиков:
средняя оценка    5422
высокая оценка    1692
низкая оценка       55
Name: critic_score_category, dtype: int64


**Распределение категорий оценок критиков
Средняя оценка (от 30 до 80, не включая 80):**

Количество игр: 11 034.

Это самая многочисленная категория, что говорит о том, что большинство игр получают средние оценки от критиков.

Высокая оценка (от 80 до 100 включительно):

Количество игр: 1 692.

Игры с высокой оценкой составляют меньшую долю, но их количество всё же значительное.

Низкая оценка (от 0 до 30, не включая 30):

Количество игр: 55.

Это самая маленькая категория, что указывает на то, что игры с очень низкими оценками критиков встречаются редко.

**Итоговый вывод**
Категории созданы корректно:

Средняя оценка: 11 034 игры.

Высокая оценка: 1 692 игры.

Низкая оценка: 55 игр.

In [35]:
# Распределение оценок по жанрам
print("Распределение оценок критиков по жанрам:")
print(df_actual.groupby('genre')['critic_score_category'].value_counts())

Распределение оценок критиков по жанрам:
genre         critic_score_category
action        средняя оценка           1288
              высокая оценка            303
              низкая оценка              16
adventure     средняя оценка            239
              высокая оценка             36
              низкая оценка               4
fighting      средняя оценка            265
              высокая оценка             93
              низкая оценка               7
misc          средняя оценка            393
              высокая оценка             73
              низкая оценка               8
platform      средняя оценка            374
              высокая оценка             82
              низкая оценка               1
puzzle        средняя оценка            182
              высокая оценка             30
              низкая оценка               2
racing        средняя оценка            495
              высокая оценка            162
              низкая оценка              10

**Анализ распределения оценок по жанрам**
1. Жанры с наибольшим количеством игр с высокой оценкой:
Action: 303 игры.

Sports: 375 игр.

Shooter: 231 игра.

Role-Playing: 178 игр.

Racing: 162 игры.

**Вывод:**

Жанры Action, Sports, Shooter и Role-Playing чаще получают высокие оценки от критиков. Это может быть связано с высоким качеством игр в этих жанрах и их популярностью среди разработчиков.

2. Жанры с наименьшим количеством игр с высокой оценкой:
Puzzle: 30 игр.

Adventure: 36 игр.

Simulation: 59 игр.

**Вывод:**

Жанры Puzzle и Adventure реже получают высокие оценки. Возможно, это связано с меньшим количеством выпускаемых игр или меньшим вниманием критиков к этим жанрам.

3. Жанры с наибольшим количеством игр с низкой оценкой:
Action: 16 игр.

Racing: 10 игр.

Fighting: 7 игр.

**Вывод:**

Даже в популярных жанрах (Action, Racing, Fighting) есть игры, которые получают низкие оценки. Это может быть связано с низким качеством или неудачными релизами.

4. Жанры с наименьшим количеством игр с низкой оценкой:
Puzzle: 2 игры.

Platform: 1 игра.

Shooter: 2 игры.

**Вывод:**

Жанры Puzzle, Platform и Shooter реже получают низкие оценки. Возможно, это связано с меньшим количеством выпускаемых игр или более высоким качеством.

Итоговый вывод
Популярные жанры:

Action, Sports, Shooter и Role-Playing чаще получают высокие оценки, но также имеют больше игр с низкими оценками.


In [36]:
# Распределение оценок по платформам
print("Распределение оценок критиков по платформам:")
print(df_actual.groupby('platform')['critic_score_category'].value_counts())

Распределение оценок критиков по платформам:
platform  critic_score_category
3DS       средняя оценка            98
          высокая оценка            17
DC        высокая оценка            11
DS        средняя оценка           618
          высокая оценка            82
          низкая оценка              8
GBA       средняя оценка           353
          высокая оценка            74
          низкая оценка              3
GC        средняя оценка           323
          высокая оценка           111
          низкая оценка              3
PC        средняя оценка           328
          высокая оценка           243
PS        средняя оценка            85
          высокая оценка            26
          низкая оценка              4
PS2       средняя оценка           994
          высокая оценка           278
          низкая оценка              3
PS3       средняя оценка           523
          высокая оценка           242
          низкая оценка              3
PS4       средняя оценка  

**Анализ распределения оценок по платформам**
1. Платформы с наибольшим количеством игр с высокой оценкой:
PS2: 278 игр.

PC: 243 игры.

X360 (Xbox 360): 243 игры.

PS3: 242 игры.

XB (Xbox): 190 игр.

**Вывод:**

Платформы PS2, PC, X360 и PS3 чаще получают высокие оценки от критиков. Это может быть связано с большим количеством качественных игр на этих платформах.

2. Платформы с наименьшим количеством игр с высокой оценкой:
DC (Dreamcast): 11 игр.

WS (WonderSwan): 0 игр.

PS4: 6 игр.

XOne (Xbox One): 3 игры.

**Вывод:**

Платформы DC, WS, PS4 и XOne реже получают высокие оценки. Возможно, это связано с меньшим количеством игр или менее популярными платформами.

3. Платформы с наибольшим количеством игр с низкой оценкой:
DS: 8 игр.

Wii: 14 игр.

X360 (Xbox 360): 9 игр.

**Вывод:**

Даже на популярных платформах (DS, Wii, X360) есть игры, которые получают низкие оценки. Это может быть связано с низким качеством или неудачными релизами.


- После категоризации данных проверьте результат: сгруппируйте данные по выделенным категориям и посчитайте количество игр в каждой категории.

**Для проверки результата категоризации данных сгруппируем данные по созданным категориям и посчитаем количество игр в каждой категории. Это можно сделать для обеих категоризаций: по оценкам пользователей (user_score_category) и по оценкам критиков (critic_score_category).**

**Группировка по категориям оценок пользователей**

In [37]:
# Группировка по категориям оценок пользователей
user_score_counts = df_actual['user_score_category'].value_counts()

# Вывод результата
print("Количество игр по категориям оценок пользователей:")
print(user_score_counts)

Количество игр по категориям оценок пользователей:
средняя оценка    10379
высокая оценка     2286
низкая оценка       116
Name: user_score_category, dtype: int64


**Группировка по категориям оценок критиков**

In [38]:
# Группировка по категориям оценок критиков
critic_score_counts = df_actual['critic_score_category'].value_counts()

# Вывод результата
print("Количество игр по категориям оценок критиков:")
print(critic_score_counts)

Количество игр по категориям оценок критиков:
средняя оценка    5422
высокая оценка    1692
низкая оценка       55
Name: critic_score_category, dtype: int64


**Итоговый вывод**
Оценки пользователей:

Средняя оценка: 10 379 игр.

Высокая оценка: 2 286 игр.

Низкая оценка: 116 игр.

Оценки критиков:

Средняя оценка: 11 034 игры.

Высокая оценка: 1 692 игры.

Низкая оценка: 55 игр.

**Анализ результатов**
Средние оценки:

Большинство игр получают средние оценки как от пользователей, так и от критиков. Это говорит о том, что большинство игр находятся в "середине" по качеству.

Высокие оценки:

Игр с высокими оценками от пользователей больше, чем с высокими оценками от критиков (2 286 против 1 692). Возможно, пользователи более лояльны, чем критики.

Низкие оценки:

Игр с низкими оценками от критиков меньше, чем с низкими оценками от пользователей (55 против 116). Это может указывать на то, что критики реже ставят очень низкие оценки.

- Выделите топ-7 платформ по количеству игр, выпущенных за весь актуальный период.

**Группировка данных по платформам
Сгруппируем данные по столбцу platform и посчитаем количество игр для каждой платформы:**

In [39]:
# Группировка по платформам и подсчёт количества игр
platform_counts = df_actual['platform'].value_counts()

# Вывод результата
print("Количество игр по платформам:")
print(platform_counts)

Количество игр по платформам:
PS2     2127
DS      2120
Wii     1275
PSP     1180
X360    1121
PS3     1087
GBA      811
XB       803
PC       766
GC       542
3DS      300
PS       274
PSV      134
WiiU      74
N64       70
DC        31
GB        27
XOne      19
PS4       16
WS         4
Name: platform, dtype: int64


**Используем метод .head(7) для выбора топ-7 платформ:**

In [40]:
# Топ-7 платформ по количеству игр
top_7_platforms = platform_counts.head(7)

# Вывод результата
print("Топ-7 платформ по количеству игр:")
print(top_7_platforms)

Топ-7 платформ по количеству игр:
PS2     2127
DS      2120
Wii     1275
PSP     1180
X360    1121
PS3     1087
GBA      811
Name: platform, dtype: int64


**Топ-7 платформ по количеству игр
PS2 (PlayStation 2):** 2127 игр.

DS (Nintendo DS): 2120 игр.

Wii (Nintendo Wii): 1275 игр.

PSP (PlayStation Portable): 1180 игр.

X360 (Xbox 360): 1121 игра.

PS3 (PlayStation 3): 1087 игр.

GBA (Game Boy Advance): 811 игр.

**Итоговый вывод**
Лидеры по количеству игр:

PS2 и DS лидируют с большим отрывом, что говорит об их популярности в этот период.

Wii, PSP, X360 и PS3 также имеют значительное количество игр, что подтверждает их успех на рынке.

GBA завершает топ-7, что подчеркивает её важность в начале 2000-х.

Данные готовы для анализа:

Эти платформы можно считать самыми популярными в период с 2000 по 2013 год.

<div class="alert alert-block alert-success">✔️
    

__Комментарий от ревьюера №1__


Отличная реализация. Молодец!

---

## 5. Итоговый вывод

В конце напишите основной вывод и отразите, какую работу проделали. Не забудьте указать описание среза данных и новых полей, которые добавили в исходный датасет.

**Итоговый вывод**
1. Проделанная работа:
Предобработка данных:

Проведена очистка данных: удалены дубликаты, обработаны пропуски, исправлены некорректные типы данных.

Приведены к единому формату названия жанров, платформ и рейтингов.

Категоризация данных:

Игры разделены на категории по оценкам пользователей (user_score_category) и критиков (critic_score_category).

Фильтрация данных:

Выделен актуальный период (2000–2013 годы) для анализа истории продаж игр.

Анализ данных:

Проведена группировка и анализ данных по жанрам, платформам и рейтингам.

Выделены топ-7 платформ по количеству выпущенных игр.

2. Описание среза данных:
Период: 2000–2013 годы.

Количество строк: 12 781.

Столбцы:

Основные: name, platform, year_of_release, genre, na_sales, eu_sales, jp_sales, other_sales, critic_score, user_score, rating.

Новые поля:

user_score_category: категория оценки пользователей (низкая, средняя, высокая).

critic_score_category: категория оценки критиков (низкая, средняя, высокая).

3. Основные результаты:
Топ-7 платформ по количеству игр:

PS2, DS, Wii, PSP, X360, PS3, GBA.

Распределение оценок:

Большинство игр получают средние оценки как от пользователей, так и от критиков.

Высокие оценки чаще получают игры на платформах PS2, PC, X360 и PS3.

Жанры:

Action, Sports, Shooter и Role-Playing чаще получают высокие оценки.

Puzzle и Adventure реже получают высокие оценки, но и низкие оценки в этих жанрах встречаются реже.

<div class="alert alert-block alert-success">✔️
    

__Комментарий от ревьюера №1__

Шикарный вывод. Молодец, что описал все твои действия (кратко) по проекту. Ведь вывод раскрывает для заказчика всю твою проделанную работу и за что он платит деньги, даже если ему не интересен весь процесс исследования, то в отчёт он заглянет обязательно. Ты молодец!
</div>

<div style="border:solid blue 3px; padding: 20px">
<div class="alert alert-block alert-success">✔️
    

__Коментарий от ревьюера №2__


В остальном всё чудно😊. Твой проект так и просится на github =)   
    
Поздравляю с успешным завершением проекта 😊👍
И желаю успехов в новых работах 😊
    
---
    
От себя хочу порекомендовать тебе отличную книгу про язык Python. Она очень классная (можно купить как новую, так и на Авито попадается. Вот ссылка на pdf этой книги)
    
https://monster-book.com/avtomatizaciya-zadach-s-python
</div>